Update docs infrastructure

tcmetzger opened the pull request:
This is a WIP PR to work on updating the docs infrastructure. This addresses several issues, including #12548.
I'll try to use this PR to track my assessment of the situation, generate tasks, and work on fixing those tasks.

The goal of this PR is to use the current versions of both Sphinx and the pydata-sphinx-theme (both of which are currently pinned). There also seems to be an issue with the autoclasstoc extension that needs to be resolved

tcmetzger committed:
Temporarily remove autoclasstoc

tcmetzger committed:
Unpin Sphinx version

tcmetzger commented:
My first goal is to build the docs with the current version of Sphinx and the theme and see what things look like.

My first step was to unpin sphinx and try to build the docs with the most recent Sphinx version (6.1.3), still using the pinned version of the theme (0.9.0). This first attempt failed with this error message:

```
  File "/home/tcm/mambaforge/envs/bkdev/lib/python3.10/site-packages/autoclasstoc/utils.py", line 165, in filter_attrs
    from sphinx.util.docstrings import extract_metadata
ImportError: cannot import name 'extract_metadata' from 'sphinx.util.docstrings' (/home/tcm/mambaforge/envs/bkdev/lib/python3.10/site-packages/sphinx/util/docstrings.py)

```

I temporarily removed autoclasstoc, which made it possible to build the docs with Sphinx 6.1.3. At first glance, the locally built docs with this setup looked good (except for the missing autoclasstocs, of course).

tcmetzger commented:
As a second step, I also unpinned the theme version. I tried building the docs with Sphinx 6.1.3 and pydata-sphinx-theme 0.13.3.

With this setup, there were several issues I encountered:

1. pydata-sphinx-theme has [moved handling of favicons to the standalone extension sphinx-favicon](https://pydata-sphinx-theme.readthedocs.io/en/stable/user_guide/branding.html#add-favicons). I added the extensions and moved updated the configuration. I'd consider this a viable solution to solve this issue.
2. the config JSON for our version switcher seems to be incompatible with the new version of the theme:
  `The version switcher "https://docs.bokeh.org/switcher.json" file is malformed at least one of the items is missing the "url" or "version" key`
  I temporarily deactivated the version switcher - this will need to be properly resolved, of course.
3. the `google_analytics_id` config parameter seems to be deprecated as well, I updated it to use the [new format](https://pydata-sphinx-theme.readthedocs.io/en/stable/user_guide/analytics.html#google-analytics). I'd consider this a viable solution to solve this issue for now - we might remove Google Analytics soon anyways.
4. Towards the end of the build, Sphinx encountered the following error:
  `Handler <function update_and_remove_templates at 0x7fd21a631bd0> for event 'html-page-context' threw an exception (exception: 'str object' has no attribute 'get')` I haven't been able to resolve this issue yet.

tcmetzger committed:
Unpin theme, update favicons/GA, disable switcher

tcmetzger commented:
Based on my attempts so far, I see the following (tentative) list of to-dos:

- [x] Fix incompatibility issue between autoclasstoc and the most recent version of Sphinx
- [x] Update switcher configuration (`switcher.json`)
- [x] Switch favicons to new favicon extension
- [x] Update Google Analytics config
- [x] Research and fix `update_and_remove_templates` exception
- [x] (most likely:) update CSS and design customizations for new theme version

@bryevdv does that match your assessment of the situation? Anything that you know off the top of your head that should go on this list?

bryevdv commented on docs/bokeh/source/docs/reference/plotting/figure.rst:
```diff
@@ -10,4 +10,3 @@ figure
     :inherited-members:
     :exclude-members: js_event_callbacks, js_property_callbacks, subscribed_events, model_class_reverse_map
 
-    .. autoclasstoc::
```
FYI can comment out (rather than remove) like
```
    .. .. autoclasstoc::
```

bryevdv commented on conda/environment-test-3.10.yml:
```diff
@@ -65,11 +65,13 @@ dependencies:
     - sphinx-copybutton
     - sphinx-design
+    - sphinx-favicon
+    - sphinx-favicon
```
dupe line

bryevdv commented:
@tcmetzger the proximate problem is that sphinx 6.1.3 is too new for the theme. I am tryin to see what earlier version of sphinx is compatible.

bryevdv commented:
@tcmetzger the proximate issue seems to be that the latest theme is not compatible with the latest sphinx, but I am still confused because the latest theme also seems incompatible even with sphinx 5.3 (what we had pinned). For reference the function in question is here:
```
py39 ❯ grip update_and_remove_templates
pydata_sphinx_theme/__init__.py
196:def update_and_remove_templates(
296:    app.connect("html-page-context", update_and_remove_templates)
```

Edit: so maybe something else is going on will try to dig deeper

bryevdv commented on docs/bokeh/source/conf.py:
```diff
@@ -184,36 +185,22 @@
+
+html_theme_options["analytics"] = {
+    "google_analytics_id": "UA-27761864-7",
 }
```
This can be added directly inside the big dict definition above

bryevdv commented:
It's blowing up tryin to render `version-switcher.html` which we don't override, I think maybe there is a configuration format change.

bryevdv commented:
@tcmetzger this change is also needed to fully disable the switcher:
```diff
-    "navbar_start": ["navbar-logo", "version-switcher"],
+    "navbar_start": ["navbar-logo"],
```
currently it is still trying to render the swticher template, bit the switcher config is commented out.

bryevdv commented:
With that change, the docs build but the output is pretty wonky

<img width="1481" alt="Screenshot 2023-04-22 at 13 12 33" src="https://user-images.githubusercontent.com/1078448/233804462-b9d2d1fd-5400-41f3-b81b-34d12303b99b.png">

bryevdv commented:
Also for now can put this in the theme options:
```
"check_switcher": False
```
to re-enable the switcher. The issue is that the block for our "latest" link does not specify a version. As long as we can have the same version in two blocks (e.g. currently "3.1.0" in both the "3.1.0" block and the "latest" block) then there should be no issue. If there is a problem with having a version repeated then it may be tricky.

tcmetzger committed:
update ga and navbar in conf.py

tcmetzger committed:
Remove dupe line

tcmetzger committed:
comment out autoclasstoc directive

tcmetzger commented:
@bryevdv Thank you, I have updated this branch. 

Seems like these three to-dos are next:

- [x] Fix incompatibility issue between autoclasstoc and the most recent version of Sphinx
- [x] Update switcher configuration (switcher.json)
- [x] update CSS and design customizations for new theme version

Now that the docs are building, I'll look into updating the CSS and our customizations!

bryevdv commented:
I've updated the `switcher.json` file on the CDN, with this change the docs look like

<img width="498" alt="Screenshot 2023-04-23 at 10 01 40" src="https://user-images.githubusercontent.com/1078448/233853801-5d53a898-5c14-4428-b699-e43ad581cfc2.png">

tcmetzger commented:
@bryevdv Thank you!

bryevdv commented:
Also FYI the docs do build with latest `autoclasstoc`, this issue is that all these things in the let column of these tables are supposed to be links but don't actually become links anymore (though they are still link color for some reason)

<img width="872" alt="Screenshot 2023-04-23 at 15 19 24" src="https://user-images.githubusercontent.com/1078448/233869217-b20f4913-44aa-441e-8d83-52a4ea4ddc11.png">


Edit: I did open an issue to see if we can get any pointers for what to look into: https://github.com/kalekundert/autoclasstoc/issues/38

tcmetzger committed:
Set check_switcher to False

tcmetzger committed:
partially re-enable autoclasstoc

tcmetzger commented:
I'm still getting the `extract_metadata` ImportError if I uncomment autoclasstoc in some of the files. 

```
ImportError: cannot import name 'extract_metadata' from 'sphinx.util.docstrings' (/home/tcm/mambaforge/envs/bkdev/lib/python3.10/site-packages/sphinx/util/docstrings.py)
```

This might just be my local env, who knows... I'll focus on the CSS first, anyways!

bryevdv commented on docs/bokeh/source/conf.py:
```diff
@@ -190,38 +192,39 @@
     "github_url": "https://github.com/bokeh/bokeh",
-    "google_analytics_id": "UA-27761864-7",
     "navbar_align": "left",
+    "navbar_start": ["navbar-logo"],
```
FYI there are two entries for `"navbar_start"` now, this is replaced on the next line with 
```
"navbar_start": ["navbar-logo", "version-switcher"],
```
I'm not sure about any build issues tho -- after the CDN update, docs build locally for me now, including with the switcher enabled.

bryevdv commented:
BTW docs seem to build ~2x as fast with `autoclasstoc` enabled. I really do think it helps organize the refguide, but maybe we decide to just punt on it after all. Or we could add a env flag to turn it off in local builds.

tcmetzger committed:
Remove duplicate entry in conf.py

tcmetzger committed:
Remove layout.html customization

tcmetzger commented:
Most of the layout issues came from our custom layout.html file. I have made some small configuration updates to hopefully make this file not needed anymore. For example, we can use `:html_theme.sidebar_secondary.remove:` in any RST files where we want to disable the right sidebar (index.rst and gallery.rst). With this change, the docs look a lot better already:

![image](https://user-images.githubusercontent.com/39711796/235254496-802f8996-7123-455f-a18b-b9b63edfb0b0.png)

tcmetzger committed:
Order options alphabetically

tcmetzger committed:
Simplify custom.css

tcmetzger commented:
I have updated custom.css and removed some things that I believe are no longer needed. The docs look mostly good now! Some things to look into some more:

- [ ] The dataframe in docs/reference/sphinxext.html#module-bokeh.sphinxext.bokeh_dataframe looks much smaller than in the currently deployed version of the docs (even though I don't think I changed any of the CSS for this element):
![image](https://user-images.githubusercontent.com/39711796/235265957-7e19eb7e-ed2c-4fb1-b585-9eeb46c86301.png)

- [x] I still have several instances of autoclasstoc commented out - if I enable them, I get the error mentioned above
- [x] There are vertical scrollbars in the left sidebar, even when there are only a few entries and no scrollbar would be necessary

I'll keep iterating over the next few days!

tcmetzger committed:
Disable rtd footer to avoid unnecessary scrollbars

tcmetzger committed:
Remove autoclasstoc

tcmetzger commented:
@bryevdv I have removed autoclasstoc, as discussed.

Do you know why the CI docs build always fails with:

```
build succeeded.

The HTML pages are in build/html.
cp: cannot stat '../../src/bokeh/server/static': No such file or directory

Build finished in 828 seconds. The HTML pages are in build/html.
Expected Google Analytics script source not found in built documentation
Error: Process completed with exit code 1.
```

I never see this error on my local env, so this might be CI-related?

bryevdv commented:
@mattpap didn't you make a recent change regarding resources?

mattpap commented:
I did, but this seems unrelated. I suppose this is the offending line in `docs/bokeh/Makefile`:
```sh
&& (cp -r ../../src/bokeh/server/static $(BUILDDIR)/html/ || true) \
```
Is the working directory different in CI? If the repository is cloned correctly, then `static/` directory must exist (due to `.keep` file) and this command will succeed regardless if the directory empty or not.

bryevdv commented:
does not appear to be in a different location 

https://github.com/bokeh/bokeh/actions/runs/4879176245/jobs/8705541777?pr=13094#step:5:11

bryevdv commented:
Also this failure is only in this PR, correct? I will try to take a closer look the next few days

tcmetzger committed:
Link landingpage hero image to gallery

tcmetzger commented:
@bryevdv it would be great if you could review this PR when you have the time - feel free to merge it when possible or let me know if you can think of any changes I should add here!

bryevdv commented:
@tcmetzger This LGTM but I don't understand the CI failure. Can you make sure this is merged/rebased on latest `branch-3.2` in case that resolves things?

tcmetzger committed:
Merge remote-tracking branch 'origin/branch-3.2' into timo/update_docs_infrastructure

tcmetzger commented:
I merged the current 3.2 branch, but it still seems to be doing the same thing...

bryevdv commented:
@mattpap this failure is happening even in jobs that succeed:

https://github.com/bokeh/bokeh/actions/runs/4906312596/jobs/8760713283#step:5:4745

which is no surprise since the CI setup deletes `src/bokeh` at the top

https://github.com/bokeh/bokeh/actions/runs/4905399279/jobs/8759181354#step:3:630

But then why is it working in other cases.

bryevdv commented:
oh, @tcmetzger this is the actual failure:

> Expected Google Analytics script source not found in built documentation

We need to disable that check, or update it.

bryevdv commented:
Here it is:

https://github.com/bokeh/bokeh/blob/e8f9b1d125262dbfe5a49c4668937e286a82b274/scripts/ci/build_docs.sh#L11-L16

mattpap commented:
> (...) this is the actual failure:

Exactly, we focused on the wrong thing. `cp` silently failed, but the error was eaten up by `|| true` (in `docs/bokeh/Makefile`).

bryevdv commented:
> Exactly, we focused on the wrong thing. cp silently failed, but the error was eaten up by || true (in docs/bokeh/Makefile).

Presumably this was my quick fix for when we started deleting the source tree in CI. Open to any better suggestions. 

The "expected GA code" message could have been more prominent, e.g. with "ERROR: " in front. For now, since we nominally plan to move off GA, I'd say we can just delete that check from `build_docs.sh` entirely. 

Edit: In any case, approved pending that one update.

bryevdv committed:
Remove GA script checks

bryevdv commented:
@tcmetzger I think I was able to edit the script in the web UI. Assuming this passes now, it is good to merge.

Edit: current failures are GH outage, It think https://www.githubstatus.com/

tcmetzger commented:
@bryevdv Thanks for making the update! The docs test is now passing, so I'll go ahead and merge!
